src/libertem/executor/pipelined.py

```diff
@@ -666,14 +666,22 @@ def _start_pool(self) -> WorkerPool:
             # FIXME: check process liveness here, too, to catch early startup
             # failures without having to run into the timeout
             for i in range(pool.size):
-                with pool.response_queue.get(timeout=self._startup_timeout) as (msg, _):
-                    if msg["type"] == "ERROR":
-                        _raise_from_msg(msg, "error on startup")
-                    if msg["type"] != "STARTUP_DONE":
-                        raise RuntimeError(
-                            f"unknown message type {msg['type']}, expected STARTUP_DONE"
-                        )
-                    span.add_event("worker startup done", {"worker_id": msg["worker_id"]})
+                try:
+                    with pool.response_queue.get(timeout=self._startup_timeout) as (msg, _):
+                        if msg["type"] == "ERROR":
+                            _raise_from_msg(msg, "error on startup")
+                        if msg["type"] != "STARTUP_DONE":
+                            raise RuntimeError(
+                                f"unknown message type {msg['type']}, expected STARTUP_DONE"
+                            )
+                        span.add_event("worker startup done", {"worker_id": msg["worker_id"]})
```

Comment:
If this were put into a function, we could have two `with` blocks and add the 'waiting for worker' warning:

```python
def _inspect_startup(msg, span):
    if msg["type"] == "ERROR":
        _raise_from_msg(msg, "error on startup")
    if msg["type"] != "STARTUP_DONE":
        raise RuntimeError(
            f"unknown message type {msg['type']}, expected STARTUP_DONE"
        )
    span.add_event("worker startup done", {"worker_id": msg["worker_id"]})

warn_time = 5.
time_to_warn = min(warn_time, self._startup_timeout)
time_after_warn = max(0., self._startup_timeout - warn_time)
try:
    with pool.response_queue.get(timeout=time_to_warn) as (msg, _):
        try:
            _inspect_startup(msg)
        except WorkerQueueEmpty:
            if time_after_warn == 0.:
                raise
            warnings.warn('Slow worker startup, please wait...', RuntimeWarning)
    with pool.response_queue.get(timeout=time_after_warn) as (msg, _):
        _inspect_startup(msg)
except WorkerQueueEmpty:
    ...
```

(I didn't test the above yet!)

Reply:
This is implemented now